### PR TITLE
Add simple intersection example 

### PR DESCRIPTION
**Pull request by masterleinad**
While helping people in the Argonne hackathon last week, I recognized that we don't have any simple examples to get people started. IMHO, a simple intersection serves as a good example so I more or less copied one of the manufactured solutions for querying the tree into an example that works without all the testing harness.

### CODE REVIEW

File: examples/simple_intersection/example_intersection.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      for (int j = 0; j < ny; ++j)
+        for (int k = 0; k < nz; ++k)
+        {
+          ArborX::Point p_lower{
+              {(i - .25) * hx, (j - .25) * hy, (k - .25) * hz}};
+          ArborX::Point p_upper{
+              {(i + .25) * hx, (j + .25) * hy, (k + .25) * hz}};
+          bounding_boxes_host[ind(i, j, k)] = {p_lower, p_upper};
+        }
+    Kokkos::deep_copy(bounding_boxes, bounding_boxes_host);
```
Comment (aprokop) [inline examples/simple_intersection/example_intersection.cpp:69]: I would move bounding box construction in a separate function.

File: examples/simple_intersection/example_intersection.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    float Lz = 100.0;
+    int nx = 11;
+    int ny = 11;
+    int nz = 11;
+    int n = nx * ny * nz;
+    float hx = Lx / (nx - 1);
+    float hy = Ly / (ny - 1);
+    float hz = Lz / (nz - 1);
+
+    std::function<int(int, int, int)> ind = [nx, ny](int i, int j, int k) {
```
Comment (aprokop) [inline examples/simple_intersection/example_intersection.cpp:51]:
```suggestion
    auto index = [nx, ny](int i, int j, int k) {
```

File: examples/simple_intersection/example_intersection.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    {
+      assert(offsets_host(i) == i);
+    }
+    assert(static_cast<int>(indices_host.size()) == n);
+    for (int i = 0; i < n; ++i)
+    {
+      assert(indices_host(i) == i);
+    }
+    std::cout << "Checking results successful." << '\n';
+  }
```
Comment (aprokop) [inline examples/simple_intersection/example_intersection.cpp:112]: Maybe replace asserts with a boolean flag, and only indicate at the end whether it failed.

File: examples/simple_intersection/example_intersection.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+// i-2  i-1  i  i+1
+//
+//  o    o   o   o   j+1
+//          ---
+//  o    o | x | o   j
+//          ---
+//  o    o   o   o   j-1
+//
+//  o    o   o   o   j-2
+//
```
Comment (aprokop) [inline examples/simple_intersection/example_intersection.cpp:29]: Add a comment that the result of each box will only intersect with itself.

**Review by aprokop — COMMENTED** [review]
Overall, looks fine. Personally, I would add more in-code comments explicitly saying things like "Build the search index", "Perform the search using constructed queries", etc. Maybe also remove the output, but it's up to you.

**Review by dalg24 — CHANGES_REQUESTED** [review]
This is not readable enough to be an example

**Comment by masterleinad** [comment]
> This is not readable enough to be an example

What are you missing? Do you want more comments?
IMHO, there is a pretty clear comment on what the example is trying to do right at the top of the file and individual steps are documented via `std::cout`.

File: examples/simple_intersection/example_intersection.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <ArborX.hpp>
+
+#include <Kokkos_Core.hpp>
+
+// Do intersection queries using the same objects for the queries as the objects
+// used in BVH construction that are located on a regular spaced
+// three-dimensional grid.
```
Comment (dalg24) [inline examples/simple_intersection/example_intersection.cpp:18]: If you do self-intersection, the grid brings nothing to the table but extra complexity

File: examples/simple_intersection/example_intersection.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    std::cout << "Filling queries." << '\n';
+    auto const n = bounding_boxes.size();
+    Kokkos::View<decltype(ArborX::intersects(ArborX::Box{})) *, MemorySpace>
+        queries("queries", n);
+    Kokkos::parallel_for(
+        "fill_queries",
+        Kokkos::RangePolicy<ExecutionSpace>(execution_space, 0, n),
+        KOKKOS_LAMBDA(int i) {
+          queries(i) = ArborX::intersects(bounding_boxes(i));
+        });
```
Comment (dalg24) [inline examples/simple_intersection/example_intersection.cpp:95]: Solution that leverage access traits is superior.  You can expect people will copy and paste from our examples so we should avoid it.

Comment (dalg24) [reply examples/simple_intersection/example_intersection.cpp:95]: Deprecating what?

Comment (masterleinad) [reply examples/simple_intersection/example_intersection.cpp]: I changed the approach to using access traits now.

File: examples/simple_intersection/example_intersection.cpp (hunk 5 above)
Comment (masterleinad) [reply examples/simple_intersection/example_intersection.cpp]: In my opinion, it's useful to have a complete example that users can easily relate to it; even if that means more code than absolutely necessary.

File: examples/simple_intersection/example_intersection.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+          boxes_host[index(i, j, k)] = {p_lower, p_upper};
+        }
+    Kokkos::deep_copy(execution_space, _boxes, boxes_host);
+  }
+
+  // Return the number of boxes.
+  KOKKOS_FUNCTION int size() const { return _boxes.size(); }
+
+  // Return the box with index i.
+  KOKKOS_FUNCTION const ArborX::Box &get_box(int i) const { return _boxes(i); }
```
Comment (aprokop) [inline examples/simple_intersection/example_intersection.cpp:76]: Does not follow camel case style and east constness.

File: examples/simple_intersection/example_intersection.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+      if (offsets_host(i) != i)
+        Kokkos::abort("Wrong entry in the offsets View!\n");
+
+    if (indices_host.size() != n)
+      Kokkos::abort("Wrong dimensions for the indices View!\n");
+    for (int i = 0; i < static_cast<int>(n); ++i)
+      if (indices_host(i) != i)
+        Kokkos::abort("Wrong entry in the indices View!\n");
+    std::cout << "Checking results successful." << '\n';
+  }
```
Comment (aprokop) [inline examples/simple_intersection/example_intersection.cpp:165]: You could probably write it more compact with `std::vector`, `std::iota` and `operator==`, but no biggie.

**Review by aprokop — APPROVED** [review]
Looks fine to me for an example.

File: README.md [hunk 9]
```diff
@@ -13,6 +13,10 @@ Using ArborX
 ------------
 The interface is described [here](https://github.com/arborx/ArborX/wiki/ArborX%3A%3ABoundingVolumeHierarchy).
 
+Examples
+--------
+Examples can be found in the [examples directory](https://github.com/arborx/ArborX/examples).
```
Comment (dalg24) [inline README.md:18]: Link does not work

**Pull request merged**